apps/omg_watcher/lib/omg_watcher/block_validator.ex

```diff
@@ -64,4 +73,93 @@ defmodule OMG.Watcher.BlockValidator do
+      # A block with a merge transaction only would have no fee transaction.
+      0 ->
+        {:error, :empty_block}
+
+      n when n > @transaction_upper_limit ->
+        {:error, :transactions_exceed_block_limit}
+
+      _ ->
+        :ok
+    end
+  end
+
+  @spec verify_no_duplicate_inputs([Transaction.Recovered.t()]) :: {:ok, map()}
+  defp verify_no_duplicate_inputs(transactions) do
+    Enum.reduce_while(transactions, {:ok, %{}}, fn tx, {:ok, input_set} ->
+      tx
+      |> Map.get(:signed_tx)
+      |> Map.get(:raw_tx)
+      |> Map.get(:inputs, [])
+      # Setting an empty array as default because fee transactions will not have an `input` key.
+      |> scan_for_duplicates({:cont, {:ok, input_set}})
+    end)
+  end
+
+  # Nested reducer executing duplicate verification logic for `verify_no_duplicate_inputs/1`
+  @spec scan_for_duplicates(Transaction.Recovered.t(), {:cont, {:ok, map()}}) :: {:cont, {:ok, map()}}
+  defp scan_for_duplicates(tx_input_set, {:cont, {:ok, input_set}}) do
+    Enum.reduce_while(tx_input_set, {:cont, {:ok, input_set}}, fn input, {:cont, {:ok, input_set}} ->
+      input_position = Utxo.Position.encode(input)
+
+      case Map.has_key?(input_set, input_position) do
+        true -> {:halt, {:halt, {:error, :block_duplicate_inputs}}}
+        false -> {:cont, {:cont, {:ok, Map.put(input_set, input_position, true)}}}
+      end
+    end)
+  end
+
+  @spec verify_fee_transactions([Transaction.Recovered.t()]) :: {:ok, [Transaction.Recovered.t()]} | {:error, atom()}
+  defp verify_fee_transactions(transactions) do
+    identified_fee_transactions = Enum.filter(transactions, &is_fee/1)
+
+    with :ok <- expected_index(transactions, identified_fee_transactions),
+         :ok <- unique_fee_transaction_per_currency(identified_fee_transactions) do
+      {:ok, identified_fee_transactions}
+    end
+  end
+
+  @spec expected_index([Transaction.Recovered.t()], [Transaction.Recovered.t()]) :: :ok | {:error, atom()}
+  defp expected_index(transactions, identified_fee_transactions) do
+    number_of_fee_txs = length(identified_fee_transactions)
+    tail = Enum.slice(transactions, -number_of_fee_txs, number_of_fee_txs)
+
+    case identified_fee_transactions do
+      ^tail -> :ok
+      _ -> {:error, :unexpected_transaction_type_at_fee_index}
+    end
+  end
+
+  @spec unique_fee_transaction_per_currency([Transaction.Recovered.t()]) :: :ok | {:error, atom()}
+  defp unique_fee_transaction_per_currency(identified_fee_transactions) do
+    identified_fee_transactions
+    |> Enum.uniq_by(fn fee_transaction -> fee_transaction |> get_fee_output() |> Map.get(:currency) end)
+    |> case do
+      ^identified_fee_transactions -> :ok
+      _ -> {:error, :duplicate_fee_transaction_for_ccy}
+    end
+  end
+
+  defp is_fee(%Transaction.Recovered{signed_tx: %Transaction.Signed{raw_tx: %Transaction.Fee{}}}),
+    do: true
+
+  defp is_fee(_), do: false
+
+  defp get_fee_output(fee_transaction) do
```

Comment:
You should be able to use `fee_transaction |> Transaction.get_outputs() |> Enum.at(0)`